File: 02_Data_Structures/linked_lists/doubly_linked_list/implementations/operations.py
# Doubly Linked List Operations (Python)
# ---------------------------------------
#
# This file implements basic operations on a Doubly Linked List:
#   - Insert at front
#   - Insert at end
#   - Delete a value
#   - Traverse forward and backward
#
# Time Complexity:
#   Insert front: O(1)
#   Insert end: O(1)
#   Delete node reference: O(1)
#   Delete by value: O(n)
#   Traversal: O(n)
#
# Space Complexity: O(1) extra
#
# A DLL node has: data, prev pointer, next pointer.
#
# Practical use cases:
#   Implementing LRU cache, text editors (undo/redo), deque-like structures.


class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    # Insert at the beginning
    def insert_front(self, value):
        new_node = Node(value)

        if self.head is None:
            self.head = self.tail = new_node
            return

        new_node.next = self.head
        self.head.prev = new_node
        self.head = new_node

    # Insert at the end
    def insert_end(self, value):
        new_node = Node(value)

        if self.tail is None:
            self.head = self.tail = new_node
            return

        self.tail.next = new_node
        new_node.prev = self.tail
        self.tail = new_node

    # Delete the first occurrence of value
    def delete_value(self, value):
        if self.head is None:
            return

        curr = self.head
        while curr and curr.data != value:
            curr = curr.next

        if curr is None:
            return  # not found

        # Case 1: delete head
        if curr == self.head:
            self.head = curr.next
            if self.head:
                self.head.prev = None
            else:
                self.tail = None  # list becomes empty

        # Case 2: delete tail
        elif curr == self.tail:
            self.tail = curr.prev
            self.tail.next = None

        # Case 3: delete middle node
        else:
            curr.prev.next = curr.next
            curr.next.prev = curr.prev

        del curr

    # Display list forward
    def display_forward(self):
        curr = self.head
        result = []
        while curr:
            result.append(curr.data)
            curr = curr.next
        return result

    # Display list backward
    def display_backward(self):
        curr = self.tail
        result = []
        while curr:
            result.append(curr.data)
            curr = curr.prev
        return result


# Example usage:
if __name__ == "__main__":
    dll = DoublyLinkedList()
    dll.insert_front(10)
    dll.insert_front(20)
    dll.insert_end(5)
    dll.delete_value(20)

    print("Forward:", dll.display_forward())
    print("Backward:", dll.display_backward())

File: 02_Data_Structures/linked_lists/doubly_linked_list/implementations/operations.cpp
// Doubly Linked List Operations Implementation
// ---------------------------------------------
//
// This file provides basic operations on a Doubly Linked List (DLL):
// - Insert at front
// - Insert at end
// - Delete a node
// - Display forward and backward
//
// A DLL contains nodes with:
//   data, prev pointer, next pointer
//
// Time Complexity:
//   Insertion at head: O(1)
//   Insertion at tail: O(1) if tail maintained, else O(n)
//   Deletion: O(1) once node is found (search is O(n))
//   Traversal: O(n)
//
// Space Complexity: O(1) extra
//
// Variations:
//   - Use sentinel nodes to simplify edge cases
//   - Maintain both head and tail for O(1) operations at both ends
//
// Good for interview demonstration: pointer manipulation skills.

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* prev;
    Node* next;
    Node(int val) : data(val), prev(NULL), next(NULL) {}
};

class DoublyLinkedList {
public:
    Node* head;
    Node* tail;

    DoublyLinkedList() {
        head = NULL;
        tail = NULL;
    }

    // Insert a node at the beginning
    void insertFront(int val) {
        Node* n = new Node(val);

        if (head == NULL) {
            head = tail = n;
            return;
        }

        n->next = head;
        head->prev = n;
        head = n;
    }

    // Insert a node at the end
    void insertEnd(int val) {
        Node* n = new Node(val);

        if (tail == NULL) {
            head = tail = n;
            return;
        }

        tail->next = n;
        n->prev = tail;
        tail = n;
    }

    // Delete first occurrence of value
    void deleteValue(int val) {
        if (head == NULL) return;

        Node* curr = head;

        while (curr && curr->data != val) {
            curr = curr->next;
        }

        if (!curr) return; // not found

        // Case 1: deleting head
        if (curr == head) {
            head = head->next;
            if (head) head->prev = NULL;
            else tail = NULL; // list became empty
        }
        // Case 2: deleting tail
        else if (curr == tail) {
            tail = tail->prev;
            tail->next = NULL;
        }
        // Case 3: deleting middle node
        else {
            curr->prev->next = curr->next;
            curr->next->prev = curr->prev;
        }

        delete curr;
    }

    // Display list forward
    void displayForward() {
        Node* temp = head;
        cout << "Forward: ";
        while (temp) {
            cout << temp->data << " ";
            temp = temp->next;
        }
        cout << endl;
    }

    // Display list backward
    void displayBackward() {
        Node* temp = tail;
        cout << "Backward: ";
        while (temp) {
            cout << temp->data << " ";
            temp = temp->prev;
        }
        cout << endl;
    }
};

int main() {
    DoublyLinkedList dll;
    dll.insertFront(10);
    dll.insertFront(20);
    dll.insertEnd(5);
    dll.deleteValue(20);

    dll.displayForward();
    dll.displayBackward();

    return 0;
}
